Implement application (#1)
ioxio-dataspace/example-app: README.md

# IOXIO Dataspace example app

This demo application is designed to show a practical example on how to create an
application using Dataspace tecnhologies.

It consists of simple Python backend that is responsible for authentication and data
retrieval and React-based frontend application.

You can check the [online demo](TODO: add a link there) or run it locally (this way
requires some configuration, so check [configuration](#configuration) section below).

> Main idea is to demostrate how to:

Reviewer: ```suggestion
Main idea is to demonstrate how to:
```

- Perform authentication in a dataspace

> - Handle application configuration correctly

Author: This is not the purpose of the application. People can handle their configuration however they want.

- Retrieve data products from Product Gateway

## Repo structure

- [backend](./backend/) - FastAPI backend
  - [main.py](./backend/app/main.py) - All the backend routes, e.g for authentication or
    data retrieval
  - [settings.py](./backend/app/settings.py) - Backend configuration
- [frontend](./frontend) - React application
  - [containers](./frontend/src/containers) - Root containers for handling data products
  - [components](./frontend/src/components) - Stateless components to simplify following
    the containers' logic
  - [utils](./frontend/src/utils) - Some helpers, e.g for making network requests to
    backend

## Local installation

### Configuration

Before running the app locally, you have to:

1. Register an application in

>    [Developer Portal](https://developer.sandbox.ioxio-dataspace.com/)

Reviewer: It would be great if this would show details on how to fill in the `Redirect URIs` and `Logout Redirect URIs`, since the paths fro those are not necessarily obvious without reading the source code in detail.

2. Obtain Client ID and Client secret of the application
3. Create [backend/.env](backend/.env) file based on
   [backend/.env.example](backend/.env.example) and set the variables with values taken
   from the previous step

### Pre-requisites

> - [Python 3.10](https://www.python.org/) - For running backend

Reviewer: pyproject.toml says python ^3.9, so either this could be laxed to also say python ^3.9 or the pyproject.toml should be updated.

# IOXIO Dataspace example app

This demo application is designed to show a practical example on how to create an

> application using Dataspace technologies.

Author: ```suggestion
application that connects to a Dataspace built with the IOXIO Dataspace technology.
```

> It consists of simple Python backend that is responsible for authentication and data

Author: Python is a programming language, React is not. Say Python FastAPI.
```suggestion
It consists of a simple Python [FastAPI](https://fastapi.tiangolo.com) backend that is responsible for authentication and data
```

> It consists of simple Python backend that is responsible for authentication and data
> retrieval and React-based frontend application.

Reviewer: ```suggestion
It consists of a simple Python backend that is responsible for authentication and data
retrieval and a React-based frontend application.
```

> You can check the [online demo](https://example-app.demos.ioxio.dev) or run it locally
> (check the [configuration](#configuration) section for the instructions).

Author: ```suggestion
You can try the [online demo](https://example-app.demos.ioxio.dev) or check the [configuration](#configuration) section for instructions on how to run this code locally.
```

Main idea is to demonstrate how to:

- Perform authentication in a dataspace

> - Use the application credentials

Reviewer: I don't get what `Use the application credentials` means. Is this something for the future when there's authentication of the app as well?

Author: Should probably say "Use the authentication tokens for data products"

> - Retrieve data products from Product Gateway

Reviewer: I'd move the `Retrieve data products from Product Gateway` to be the first one, as it's the most important part in my opinion.

## Repo structure

> - [backend](./backend/) - FastAPI backend

Reviewer: ```suggestion
- [backend](./backend/) - Python [FastAPI](https://fastapi.tiangolo.com/) backend
```

>   - [main.py](./backend/app/main.py) - All the backend routes, e.g for authentication or
>     data retrieval

Reviewer: ```suggestion
  - [main.py](./backend/app/main.py) - All the backend routes, e.g. for authentication 
    or data retrieval
```

Author: It's not "or", it's "and".

  - [settings.py](./backend/app/settings.py) - Backend configuration
- [frontend](./frontend) - React application
  - [containers](./frontend/src/containers) - Root containers for handling data products
  - [components](./frontend/src/components) - Stateless components to simplify following
    the containers' logic

>   - [utils](./frontend/src/utils) - Some helpers, e.g for making network requests to
>     backend

Reviewer: ```suggestion
  - [utils](./frontend/src/utils) - Some helpers, e.g. for making network requests to
    the backend
```

## Local installation

### Configuration

Before running the app locally, you have to:

> 1. Register an application in
>    [Developer Portal](https://developer.sandbox.ioxio-dataspace.com/). Use the following
>    values in the form:
>    - Redirect URI: `http://localhost:3000/api/auth`
>    - Logout redirect URI: `http://localhost:3000`
> 2. Obtain Client ID and Client secret of the application
> 3. Create [backend/.env](backend/.env) file based on
>    [backend/.env.example](backend/.env.example) and set the variables with values taken
>    from the previous step

Reviewer: Just a suggestion; feel free to use any parts you like or adapt (or leave as it is if you prefer that)

```suggestion
1. Register your application in the
   [Developer Portal](https://developer.sandbox.ioxio-dataspace.com/). Use the following
   values in the form:
   - Redirect URI: `http://localhost:3000/api/auth`
   - Logout redirect URI: `http://localhost:3000`

   Note: In the next step you will need the Client ID and Client secret that get 
   generated when you complete the registration of the application.

2. Create the [backend/.env](backend/.env) file based on
   [backend/.env.example](backend/.env.example) and set the variables with the values 
   from the previous step.
```

Author: `+` "Please keep in mind that when deploying your ready application, you will need to update these URIs to point to your deployment."

### Pre-requisites

> - [Python 3.9+](https://www.python.org/) - For running backend

Author: ```suggestion
- [Python 3.9+](https://www.python.org/) - For running the backend
```

> - [Poetry](https://python-poetry.org/) - Python package manager

Author: `NPM` = `Node Package Manager`. `Poetry` != `Python Package Manager`
```suggestion
- [Poetry](https://python-poetry.org/) - Python dependency management tool
```

> - [Node](https://nodejs.org/en/) - For running frontend

Author: ```suggestion
- [Node](https://nodejs.org/en/) - For running the frontend
```

- [pnpm](https://pnpm.io/) - Node package manager

> - (Optionally) [pre-commit](https://pre-commit.com/) - If you want to contribute to this
>   repo, these hooks will be executed before a commit. Don't forget to run
>   `pre-commit install` in this case after checking out the repo.

Author: ```suggestion
- [pre-commit](https://pre-commit.com/) - Runs hooks before you commit to e.g. format your code. Make sure you run `pre-commit install` after checking out the repo.
```

Author: We can tell people it's a good idea to use pre-commit in general, not only when contributing back to this repo.

### Backend

```bash
cd backend
poetry install

> # run it:

Author: ```suggestion

```

Author: GitHub is showing the suggestion wrong, I'm suggesting replacing it with an empty line.

poetry run dev
```

### Frontend

```bash
cd frontend
pnpm install

> # run it:

Author: ```suggestion

```

Author: GitHub is showing the suggestion wrong, I'm suggesting replacing it with an empty line.
